File: Java/1029.two-city-scheduling.java
/*
 * @lc app=leetcode id=1029 lang=java
 *
 * [1029] Two City Scheduling
 */

// @lc code=start
class Solution {
    /*
    按照A[0] - A[1]排序，A[0] - A[1]差值大的按greedy被分配对应值小的
    eg: [30, 200] -> -170, 放入A中, [200, 30] -> 170 放入B中
    time: O(nlogn)
    space: O(1)
    */
    public int twoCitySchedCost(int[][] costs) {
        Arrays.sort(costs, (a, b) -> (a[0] - a[1] - (b[0] - b[1])));
        int res = 0;
        for (int i = 0; i < costs.length; ++i) {
            res += i < costs.length / 2 ? costs[i][0] : costs[i][1];
        }
        return res;
    }
}
// @lc code=end

File: Cpp/1029.two-city-scheduling.cpp
/*
 * @lc app=leetcode id=1029 lang=cpp
 *
 * [1029] Two City Scheduling
 */

// @lc code=start
class Solution {
/*
dp[i][j]: 前i个人中，有j个人在A的最小cost
dp[i][j] = min(dp[i - 1][j - 1] + costs[i - 1][0], dp[i - 1][j] + costs[i - 1][1])
注意初始化dp[i][0]为dp[i - 1][0] + costs[i - 1][1]
dp[0][0] = 0
time: O(n^2)
space: O(n^2)
*/
public:
    int twoCitySchedCost(vector<vector<int>>& costs) {
        int n = costs.size();
        vector<vector<int>> dp(n + 1, vector<int>(n + 1, INT_MAX / 2)); // INT_MAX will out of boundary
        dp[0][0] = 0;
        for (int i = 1; i <= n; ++i) {
            dp[i][0] = dp[i - 1][0] + costs[i - 1][1];
            for (int j = 1; j <= i; ++j) {
                dp[i][j] = min(dp[i - 1][j - 1] + costs[i - 1][0],
                               dp[i - 1][j] + costs[i - 1][1]);
            }
        }
        return dp[n][n / 2];
    }
};
// @lc code=end

